### ISSUE TITLE
Caching data of chat_analyzer

### ISSUE THREAD

**Issue by ZsemberiDaniel** [ZsemberiDaniel/facebook_msg_parser#13]
Adding cached data to chat_analyzer functions would maybe be a good idea, so we don't have to calculate that each time a function is called, because these are the most demanding functions of the application.

Function to consider:
- [x] photo_count_per_participant (used in titles)
- [x] gif_count_per_participant (used in titles)
- [x] share_count_per_participant (used in titles)
- [x] message_count (used by 1 other and basic)
- [ ] message_by_day (used by 1 other)
- [ ] message_count_monthly (used by titles)
- [ ] _response_times_sum_by_day (used by 2 other)
- [ ] avg_response_time (used by basic and titles)
- [x] response_count (used by 2 other and basic)
- [ ] character_count_per_participant (used by 2 other and basic)
- [ ] character_by_day (used by 1 other)
- [x] a bool for the message whether it has an emoji or not

**Comment by ZsemberiDaniel** [comment]
I feel like this is a big enhancement, and should speed up the app quite a bit. So maybe this will be the only issue in this milestone, because this needs to be worked on quite a bit.

There are two ways to do it. It can be done the first time a function is called.
* Has the advantage of not calculating unnecessary information.
* The user does not have to wait at the start of the application for it to do all the calculations.
* The first time a function is called it just saves the data and next time reads it.
The other way is to calculate everything at the start of the application.
* Everything is calculated beforehand -> really fast application.
* The user has to wait a long time at the start.

So the question comes down to whether we want the user to wait a long time at the start, or wait a shorter time at the first call of all command, and I think the second one seems to be better, because a long preparation time might scare away the user.

**Comment by ZsemberiDaniel** [comment]
So after I stopped being a dumbass I started a profiler and actually went ahead and looked at what might be causing the slow running times of the commands basic, titles and etc.

Turns out it's not the counting algorithms but rather, the emoji searching algorithms, especially regex. Yeah, like usually the 100x slow down is regex's fault (ok, I'm being a bit too harsh on regex here).

So I solved that by implementing the regex search once, while deconding the string and then storing that to the decoded_message.json.

The caching is already implemented, and is being used on some functions, but that's only a 2% speed gain, so I might eliminate that completely, and maybe leave the caching function in there just in case I want to use it in the future.
